src/cellars.rs

```diff
@@ -86,6 +224,113 @@ pub fn normalize_address(address: String) -> String {
 
 // validation logic
 
+pub fn validate_new_adaptor(
+    cellar_id: &str,
+    adaptor_id: &str,
+    permissions: &CellarArchPermissions,
+) -> Result<(), Error> {
+    let adaptor_id = normalize_address(adaptor_id.to_string());
+    check_blocked_adaptor(&adaptor_id)?;
+
+    let cellar_id = normalize_address(cellar_id.to_string());
+    if !permissions
+        .allowed_adaptors
+        .contains(&(&cellar_id, &adaptor_id))
+    {
+        return Err(sp_call_error(format!(
+            "new adaptor {adaptor_id} not allowed to be added for cellar {cellar_id}"
+        )));
+    }
+
+    Ok(())
+}
+
+pub fn validate_new_position(
+    cellar_id: &str,
+    position: u32,
+    permissions: &CellarArchPermissions,
+) -> Result<(), Error> {
+    check_blocked_position(&position, permissions)?;
+
+    let cellar_id = normalize_address(cellar_id.to_string());
+    if !permissions
+        .allowed_positions
+        .contains(&(&cellar_id, position))
+    {
+        return Err(sp_call_error(format!(
+            "new position {position} not allowed to be added for cellar {cellar_id}"
+        )));
+    }
+
+    Ok(())
+}
+
+pub fn validate_oracle(
+    cellar_id: &str,
+    registry_id_in: &str,
+    oracle_in: &str,
+) -> Result<(), Error> {
+    let cellar_id_normalized = normalize_address(cellar_id.to_string());
+    let oracle_in = normalize_address(oracle_in.to_string());
+    let registry_id_in = string_to_u256(registry_id_in.to_string())?;
+    if cellar_id_normalized.eq(CELLAR_TURBO_SWETH)
+        && ALLOWED_TURBOSWETH_PRICE_ORACLES.contains(&(registry_id_in, oracle_in.as_str()))
+    {
+        return Ok(());
+    }
+    if cellar_id_normalized.eq(CELLAR_TURBO_SOMM)
+        && ALLOWED_TURBOSOMM_PRICE_ORACLES.contains(&(registry_id_in, oracle_in.as_str()))
+    {
+        return Ok(());
+    }
+
+    Err(sp_call_error("unauthorized oracle update".to_string()))
+}
+
+pub fn validate_cache_price_router(
+    cellar_id: &str,
+    check_total_assets_value: bool,
+    allowable_range_value: u32,
+    expected_price_router: Option<String>,
+) -> Result<(), Error> {
+    let cellar_id_normalized = normalize_address(cellar_id.to_string());
+    let expected_price_router = expected_price_router.map(normalize_address);
+
+    if !ALLOWED_CACHE_PRICE_ROUTER
+        .iter()
+        .any(|(cellar_id, permissions)| {
+            cellar_id_normalized.eq(cellar_id)
+                && permissions.0 == check_total_assets_value
+                && permissions.1 >= allowable_range_value
+                && permissions.2 == expected_price_router
+        })
+    {
+        return Err(sp_call_error("call not authorized for cellar".to_string()));
+    }
+
+    Ok(())
+}
+
+pub fn validate_force_position_out(
+    cellar_id: &str,
+    index: u32,
+    position_id: u32,
+    in_debt_array: bool,
+) -> Result<(), Error> {
+    let cellar_id_normalized = normalize_address(cellar_id.to_string());
```

Comment:
We may actually be able to remove this as I assume this was a one-time fix, and we can move `ForcePositionOut` to gov.

Reply:
Ah responded to your earlier comment before seeing this. Yes I think it was one-off. will fix.